packages/contracts-bedrock/scripts/L2Genesis.s.sol

```diff
@@ -0,0 +1,345 @@
+    function _setOptimismMintableERC20Factory() internal {
+        address impl = _predeployToCodeNamespace(Predeploys.OPTIMISM_MINTABLE_ERC20_FACTORY);
+        OptimismMintableERC20Factory factory = new OptimismMintableERC20Factory();
+
+        vm.etch(impl, address(factory).code);
+        vm.store(
+            Predeploys.OPTIMISM_MINTABLE_ERC20_FACTORY,
+            // uint8 private _initialized;
+            // bool private _initializing;
+            hex"0000000000000000000000000000000000000000000000000000000000000000",
+            // 1
+            // false
+            hex"0000000000000000000000000000000000000000000000000000000000000001"
+        );
+        vm.store(
+            Predeploys.OPTIMISM_MINTABLE_ERC20_FACTORY,
+            // address public bridge;
+            hex"0000000000000000000000000000000000000000000000000000000000000001",
+            // L2_STANDARD_BRIDGE
+            hex"0000000000000000000000004200000000000000000000000000000000000010"
+        );
+
+        vm.etch(address(factory), hex"");
+        vm.resetNonce(address(factory));
+    }
+
+    function _setL1BlockNumber() internal {
+        vm.etch(
+            _predeployToCodeNamespace(Predeploys.L1_BLOCK_NUMBER),
+            vm.getDeployedCode("L1BlockNumber.sol:L1BlockNumber")
+        );
+    }
+
+    function _setGasPriceOracle() internal {
+        vm.etch(
+            _predeployToCodeNamespace(Predeploys.GAS_PRICE_ORACLE),
+            vm.getDeployedCode("GasPriceOracle.sol:GasPriceOracle")
+        );
+    }
+
+    function _setGovernanceToken() internal {
+        if (!cfg.enableGovernance()) {
+            console.log("Governance not enabled, skipping setting governanace token");
+            return;
+        }
+
+        vm.etch(Predeploys.GOVERNANCE_TOKEN, vm.getDeployedCode("GovernanceToken.sol:GovernanceToken"));
+
+        vm.store(
+            Predeploys.GOVERNANCE_TOKEN,
+            // string private _name;
+            hex"0000000000000000000000000000000000000000000000000000000000000003",
+            // "Optimism"
+            hex"4f7074696d69736d000000000000000000000000000000000000000000000010"
+        );
+        vm.store(
+            Predeploys.GOVERNANCE_TOKEN,
+            // string private _symbol;
+            hex"0000000000000000000000000000000000000000000000000000000000000004",
+            // "OP"
+            hex"4f50000000000000000000000000000000000000000000000000000000000004"
+        );
+        vm.store(
+            Predeploys.GOVERNANCE_TOKEN,
+            // address private _owner;
+            hex"000000000000000000000000000000000000000000000000000000000000000a",
+            bytes32(uint256(uint160(cfg.governanceTokenOwner())))
+        );
+    }
+
+    function _setL1Block() internal {
+        vm.etch(
+            _predeployToCodeNamespace(Predeploys.L1_BLOCK_ATTRIBUTES),
+            vm.getDeployedCode("L1Block.sol:L1Block")
+        );
+
+        // TODO - These attributes were set in previous L2 genesis,
```

Comment:
Can you clarify this comment?

Reply:
Current L2 genesis has [these](https://github.com/ethereum-optimism/optimism/blob/55a30f48b855dbcfb4cf77dac7b1864b026c20cd/op-chain-ops/genesis/config.go#L865-L874) values set initially for the `L1Block` predeploy

For the following values, should I source them from the `genesis-l1.json` file?
```go
storage["L1Block"] = state.StorageValues{
		"number":         block.Number(),
		"timestamp":      block.Time(),
		"basefee":        block.BaseFee(),
		"hash":           block.Hash(),
	}
```

Reply:
I've decided to just add these properties to the deploy config. For spinning up the local devnet, I figured we'd be able to add these `block.` properties to the deploy config after creating the L1 genesis

```solidity
function _setL1Block() internal {
        address impl = _predeployToCodeNamespace(Predeploys.L1_BLOCK_ATTRIBUTES);
        vm.etch(
            impl,
            vm.getDeployedCode("L1Block.sol:L1Block")
        );

        vm.startPrank(0xDeaDDEaDDeAdDeAdDEAdDEaddeAddEAdDEAd0001);
        L1Block(Predeploys.L1_BLOCK_ATTRIBUTES).setL1BlockValues(
            uint64(cfg.l1GenesisBlockNumber()),
            uint64(cfg.l1GenesisBlockTimestamp()),
            uint64(cfg.l1GenesisBaseFee()),
            cfg.l1GenesisBlockHash(),
            uint64(cfg.l1GenesisSequenceNumber()),
            bytes32(uint256(uint160(cfg.batchSenderAddress()))),
            cfg.gasPriceOracleOverhead(),
            cfg.gasPriceOracleScalar()
        );
        vm.stopPrank();

        _verifyProxyImplementationAddress(Predeploys.L1_BLOCK_ATTRIBUTES, impl);
    }
```

Reply:
The problem is that now its much more difficult to configure the system as all of those values must be consistent with each other. I recommend we just leave the storage slots empty as block 1 will set them to the proper values. Having them set in genesis only makes the diffs between genesis json files for different chains larger

Reply:
I'm very much for leaving these slots empty! Just didn't know if that was acceptable or not given that we are currently setting them

Reply:
[Updated](https://github.com/ethereum-optimism/optimism/pull/9292/files#diff-2b05d0be8d7d8de7870ac51225a0728adeb9a6df9819eaa654c455a23b8f509eR262-R271)!

Reply:
Thank you!